Civi/Financeextras/Event/Listener/ContributionPaymentUpdatedListener.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace Civi\Financeextras\Event\Listener;
+
+use Civi\Financeextras\Event\ContributionPaymentUpdatedEvent;
+
+class ContributionPaymentUpdatedListener {
+
+  public static function handle(ContributionPaymentUpdatedEvent $event) {
+    self::updateContributionStatus($event->getContributionId());
+  }
+
+  private static function updateContributionStatus($contributionId) {
+    $status = 'Pending';
+    $contribution = \Civi\Api4\Contribution::get()
+      ->addWhere('id', '=', $contributionId)
+      ->execute()
+      ->first();
+
+    if (empty($contribution)) {
+      return;
+    }
+
+    $status = match (TRUE) {
+      empty($contribution['paid_amount']) => 'Pending',
+      $contribution['total_amount'] == $contribution['paid_amount'] => 'Completed',
+      $contribution['total_amount'] > $contribution['paid_amount'] => 'Partially paid',
+      $contribution['total_amount'] < $contribution['paid_amount'] => 'Pending refund',
+    };
+
+    \Civi\Api4\Contribution::update()
```

Comment:
While it is easier to do it through the API, the problem that it will result in emitting the pre and post hooks which is not desired  here, so I will instead updating the status using the Contribution DAO->save() method, this means you will need to convert the status to its corresponding value, which is more work here but more efficient compared to the API approach.